File: crawler/src/crawler.py
#seed url
#download the html of tha page that this url leads to
#extract all urls that are present on that html page
#store if they are not already craweled and repeate steps until no new urls are found
import logging
from urllib.parse import urljoin
import requests
from bs4 import BeautifulSoup
from collections import deque
import threading

logging.basicConfig(
    format='%(asctime)s %(levelname)-8s %(message)s',
    level=logging.INFO,
    datefmt='%Y-%m-%d %H:%M:%S'
)

class Crawler:
    def __init__(self, seed_url):
        self.seed_url = seed_url
        self.url_queue = deque()
        self.visited_urls = []
        for url in seed_url:
            self.url_queue.append(url)
    
    #must download the content on the page in order to get the links from it
    def download_url(self, url): 
        return requests.get(url).text
    
    def get_linked_urls(self, url, html):
        soup = BeautifulSoup(html, 'html.parser')
        for link in soup.find_all('a'):
            path = link.get('href')
            if path and path.startswith('/'):
                path = urljoin(url, path)
            yield path
    
    def add_url_to_visit(self, url):
        if url not in self.visited_urls and url not in self.url_queue:
            return url

    def crawl(self, url_to_crawl):
        html = self.download_url(url_to_crawl)
        url_list = []
        for url in self.get_linked_urls(url_to_crawl, html):
            curr_url = self.add_url_to_visit(url)
            if(curr_url):
                url_list.append(curr_url)
        
        return url_list


    def crawl_all_urls(self):
        while self.url_queue:
            url_to_crawl = self.url_queue.popleft()
            print(url_to_crawl)
            links_to_crawl = self.crawl(url_to_crawl)
            for link in links_to_crawl:
                self.url_queue.append(link)
                self.visited_urls.append(url_to_crawl)



if __name__ == '__main__':
    Crawler(seed_url=['https://go101.org/article/channel-closing.html']).crawl_all_urls()

File: crawler/src/crawler.go
package main

import (
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
	"sync"

	"github.com/PuerkitoBio/goquery"
)

//next: push code so far to github
//maybe add a readme to github
//read the article about the distributed crawler and potentially add more features to it

type Crawler struct {
	seedUrls    []string
	urlQueue    []string
	visitedUrls map[string]bool //maps alone are not safe for concurrent use, so this cannot
	mutex       sync.Mutex      // Protects shared resources
	wg          sync.WaitGroup  // Wait for all goroutines to finish
	urlChan     chan string
	stopped     bool
}

// equivalent to a class constructor
func NewCrawler(seedUrls []string) *Crawler {
	return &Crawler{
		seedUrls:    seedUrls,
		urlQueue:    make([]string, len(seedUrls)),
		visitedUrls: make(map[string]bool),
		urlChan:     make(chan string, 100),
		stopped:     false,
	}
}

// Download the content of the page
// inside the first parens is a method receiver: specifies the type that the function can be called on, similar to having a class and functions which can then only be called on an instance of that class, i.e. self
// second set of parens is parameters, third set has required return types, both are required, nil will be returned in the case of no error
func (c *Crawler) downloadURL(pageURL string) (string, error) {
	resp, err := http.Get(pageURL)
	if err != nil {
		fmt.Println(err)
		return "", err
	}
	defer func() {
		resp.Body.Close()
	}()
	// runs when the function exits, ensures that the response body is closed, released resources, whether or not there is an error
	//deferring is usually for external resources

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to fetch page: %s", pageURL)
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return "", err
	}

	html, err := doc.Html() //HTML is a method that belongs to the goquery.Document type
	if err != nil {
		return "", err
	}

	return html, nil
}

// Extract linked URLs from the page
func (c *Crawler) getLinkedURLs(pageURL string, html string) []string {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		log.Fatal(err)
	}

	var urls []string
	baseURL, err := url.Parse(pageURL)
	if err != nil {
		log.Fatal(err)
	}

	doc.Find("a").Each(func(i int, s *goquery.Selection) { //more efficient to use a pointer in this case, it's common practice to use pointers for types of data that represent collections, data structures or anything that requires modifications or has as significant amount of data
		href, exists := s.Attr("href")
		if exists {
			parsedURL, err := url.Parse(href)
			if err == nil {
				absURL := baseURL.ResolveReference(parsedURL).String()
				urls = append(urls, absURL)
			}
		}
	})
	return urls
}

// Check if URL has already been visited or queued
func (c *Crawler) addURLToVisit(u string) bool {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if !c.visitedUrls[u] {
		c.visitedUrls[u] = true
		return true
	}
	return false
}

// Crawl the provided URL concurrently
// : This is a channel of type string, used for sending newly discovered URLs back to the main goroutine or another part of the program.
func (c *Crawler) crawl(pageURL string) {
	c.wg.Add(1)
	fmt.Println("added")
	defer func() {
		c.wg.Done()
		fmt.Println("done")
	}()

	html, err := c.downloadURL(pageURL)
	if err != nil {
		log.Println(err)
		return
	}

	linkedURLs := c.getLinkedURLs(pageURL, html)
	for _, link := range linkedURLs {
		if c.addURLToVisit(link) {
			c.urlChan <- link // Send newly discovered URLs to the channel
		}
	}
}

func (c *Crawler) crawlAllURLs() {
	// Seed the queue with initial URLs
	for _, seedURL := range c.seedUrls {
		c.urlQueue = append(c.urlQueue, seedURL)

		go c.crawl(seedURL)
	}

	go func() {
		defer func() {

		}()
	}()

	// Start a separate goroutine to handle crawling of discovered URLs
	for url := range c.urlChan {
		fmt.Println("Discovered:", url)

		go c.crawl(url) // Crawl discovered URLs concurrently
		// Wait for all crawlers to finish before closing the channel
		go func() {
			c.wg.Wait() // Wait for all crawling tasks to finish

			//dont close a closed channel
			c.mutex.Lock()
			if !c.stopped {
				close(c.urlChan)
				c.stopped = true
			}
			c.mutex.Unlock()
		}()
	}

}

func main() {
	crawler := NewCrawler([]string{"https://stackoverflow.com/questions/70908948/channel-hangs-probably-not-closing-at-the-right-place"})
	crawler.crawlAllURLs()
}
